src/ATen/native/xpu/sycl/LinearInt4.cpp

```diff
@@ -0,0 +1,235 @@
+      int constexpr TileK2 = 8;
+      int constexpr GroupK2 = SgSize * TileK2;
+      int k_body2 = padto_le(k, GroupK2 * Unroll);
+      int g_idx = it.get_group(0);
+      auto sg = it.get_sub_group();
+      int sg_id = sg.get_local_id()[0];
+      int g_n = g_idx;
+      auto sptr = ScaleAndZeros + g_n * ldb * 2;
+      auto bptr = B + g_n * k / 2;
+      auto aptr = A;
+      auto cptr = C + g_n;
+      sycl::float2 tmpAcc = {0.f, 0.f};
+      int i = 0;
+      for (; i < k_body; i += GroupK * Unroll) {
+#pragma unroll
+        for (int iu = 0; iu < Unroll; iu++) {
+          uint8_t tmps8[TileK / 2];
+          *(sycl::vec<uint8_t, TileK / 2>*)tmps8 =
+              *(sycl::vec<uint8_t, TileK / 2>*)(bptr + sg_id * TileK / 2);
+          scalar_t scale = *(sptr + sg_id * (TileK / blocksize) * 2);
+          scalar_t zero_point = *(sptr + sg_id * (TileK / blocksize) * 2 + 1);
+#pragma unroll
+          for (int ikk = 0; ikk < TileK; ikk += 2) {
+            scalarx2_t tmpA = *(scalarx2_t*)(aptr + sg_id * TileK + ikk);
+            scalarx2_t tmpB = {
+                static_cast<int8_t>((tmps8[ikk / 2] & 0x0f) - 8),
+                static_cast<int8_t>((tmps8[ikk / 2] >> 4) - 8)};
+            auto tmpAmulB = tmpA * (tmpB * scale + zero_point);
+            tmpAcc += {tmpAmulB[0], tmpAmulB[1]};
+          }
+          sptr += (GroupK / blocksize) * 2;
+          aptr += GroupK;
+          bptr += GroupK / 2;
+        }
+      }
+      if (i + GroupK2 * Unroll < k_body2) {
+        for (; i < k_body2; i += GroupK2 * Unroll) {
+#pragma unroll
+          for (int iu = 0; iu < Unroll; iu++) {
+            uint8_t tmps8[TileK2 / 2];
+            *(sycl::vec<uint8_t, TileK2 / 2>*)tmps8 =
+                *(sycl::vec<uint8_t, TileK2 / 2>*)(bptr + sg_id * TileK2 / 2);
+            scalar_t scale = *(sptr + sg_id * (TileK2 / blocksize) * 2);
+            scalar_t zero_point =
+                *(sptr + sg_id * (TileK2 / blocksize) * 2 + 1);
+#pragma unroll
+            for (int ikk = 0; ikk < TileK2; ikk += 2) {
+              scalarx2_t tmpA = *(scalarx2_t*)(aptr + sg_id * TileK2 + ikk);
+              scalarx2_t tmpB = {
+                  static_cast<int8_t>((tmps8[ikk / 2] & 0x0f) - 8),
+                  static_cast<int8_t>((tmps8[ikk / 2] >> 4) - 8)};
+              auto tmpAmulB = tmpA * (tmpB * scale + zero_point);
+              tmpAcc += {tmpAmulB[0], tmpAmulB[1]};
+            }
+            sptr += (GroupK2 / blocksize) * 2;
+            aptr += GroupK2;
+            bptr += GroupK2 / 2;
+          }
+        }
+      }
+      if (i + SgSize * 2 <= k) {
+        for (; i < k; i += SgSize * 2) {
+          uint8_t tmps8 = *(bptr + sg_id);
+          scalarx2_t tmpB = {
+              static_cast<int8_t>((tmps8 & 0x0f) - 8),
+              static_cast<int8_t>((tmps8 >> 4) - 8)};
+          scalar_t scale = *(sptr + (sg_id * 2 / blocksize) * 2);
+          scalar_t zero_point = *(sptr + (sg_id * 2 / blocksize) * 2 + 1);
```

Comment:
scales and zero points are always interleaving? I prefer to use a different zero_points pointer to make it compatible with different data types of zero points.

```suggestion
          scalar_t scale = *(sptr + SCALE_OFFSET);
          scalar_t zero_point = *(zptr+ZP_OFFSET);
```

Reply:
https://github.com/intel/torch-xpu-ops/pull/1130/commits/f44ed700d1c5845243f323f9a9f590fd4dd1baf1